### PR TITLE
Use explicit promotion for constants in repeat expressions

### PR DESCRIPTION
**Pull request by oli-obk**
As per https://github.com/rust-lang/rust/issues/49147#issuecomment-521210460
> Even in generics everything is obvious. Either you have a Copy bound, or you get promotion. I think not being a full const context will confuse more than it will not. We'll just get requests to add more things or at least issues opened asking why things that are obv OK don't work

cc @rust-lang/wg-const-eval

### CODE REVIEW

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -3399,7 +3399,9 @@ impl<'test> TestCx<'test> {
 
         for output_file in &files {
             if actual.is_empty() {
-                self.delete_file(output_file);
+                if output_file.exists() {
+                    self.delete_file(output_file);
+                }
```
Comment (RalfJung) [inline src/tools/compiletest/src/runtest.rs]: This is an unrelated change, right? Seems reasonable though. (Still technically a race condition though, would be safer to just "delete and ignore errors", but whatever.)

**Comment by RalfJung** [comment]
Uh, I don't actually understand most of the code you are touching here -- the glue code that connects const-eval with the rest of the compiler. I could rubber-stamp it as it looks simple enough, but maybe we can think of someone who can do a meaningful review? @eddyb? (They have a lot on their plate though.)

However, taking a step back... this is actually a pretty big decision, I feel. Last time we talked, it was still the case that rustc *first* checked if the repeat expression could be promoted, and only if that failed, it checked the `Copy` trait. Has that changed since then? Is there a way to test this? It is *crucial* that we check the `Copy` trait first, to maintain the principle that *with explicit promotion rules, failure to promote must imply failure to compile*. I wonder if there is a `delay_span_bug` we can put somewhere to enshrine this principle in code?

And even with that principle upheld, there is still the issue that it might be hard to predict whether this code runs at run-time or at compile-time (because that depends on whether the `Copy` bound is satisfied). At some point there should be a T-lang FCP to decide between these two options:
* What the PR implements: the absence of the `Copy` bound is sufficient to trigger compile-time evaluation.
* An alternative: demand some explicit marker to request compile-time evaluation, such as `[const { $expr }; N]`.

That doesn't have to happen here, it could be part of a stabilization FCP, but it should come up explicitly.

File: src/tools/compiletest/src/runtest.rs (hunk 1 above)
Comment (oli-obk) [reply src/tools/compiletest/src/runtest.rs]: Yea, I didn't want to touch the `delete_file` function, but I'll do that now

**Comment by oli-obk** [comment]
> An alternative: demand some explicit marker to request compile-time evaluation, such as [const { $expr }; N].

I am working on a prototype for this since a few days, it seems like something we want independently of this PR

**Comment by oli-obk** [comment]
> It was still the case that rustc first checked if the repeat expression could be promoted, and only if that failed, it checked the Copy trait. Has that changed since then?

This is exactly what this PR does (and what it was doing with implicit promo. If the type is `!Copy`, we mark the repeat element as a promotion candidate.

> Is there a way to test this?

I believe we are already testing it, but I can add comments to those tests so if some day in the future something changes there, ppl know what we want to happen.

> It is crucial that we check the Copy trait first, to maintain the principle that with explicit promotion rules, failure to promote must imply failure to compile. I wonder if there is a delay_span_bug we can put somewhere to enshrine this principle in code?

We can probably make the codegen code around `Rvalue::Repeat` ICE in case its element is a non-item constant whose type is `Copy`.

**Comment by RalfJung** [comment]
> This is exactly what this PR does

This PR doesn't change the order in which `Copy`-check and promotion-check are done, does it? Where is that order determined:
Back then I was first told `Copy` would be checked first but later @eddyb said that was a mistake and actually promotion was checked first.

> We can probably make the codegen code around Rvalue::Repeat ICE in case its element is a non-item constant whose type is Copy.

I feel it would be more direct and more local to have a `delay_span_bug` when promotion fails and `Validator::explicit` is `true`.

File: src/librustc_mir/transform/promote_consts.rs [hunk 2]
```diff
@@ -200,14 +202,18 @@ impl<'tcx> Visitor<'tcx> for Collector<'_, 'tcx> {
...
+        match rvalue {
             Rvalue::Ref(..) => {
                 self.candidates.push(Candidate::Ref(location));
             }
-            Rvalue::Repeat(..) if self.tcx.features().const_in_array_repeat_expressions => {
-                // FIXME(#49147) only promote the element when it isn't `Copy`
-                // (so that code that can copy it at runtime is unaffected).
-                self.candidates.push(Candidate::Repeat(location));
+            Rvalue::Repeat(elem, _) if self.tcx.features().const_in_array_repeat_expressions => {
+                if !elem.ty(&self.body.local_decls, self.tcx).is_copy_modulo_regions(
```
Comment (oli-obk) [inline src/librustc_mir/transform/promote_consts.rs]: > This PR doesn't change the order in which Copy-check and promotion-check are done, does it? Where is that order determined?

It only performs the promotion check if the type is not `Copy`

**Comment by oli-obk** [comment]
> I feel it would be more direct and more local to have a delay_span_bug when promotion fails and Validator::explicit is true.

Oh that makes a lot of sense. Let's do that (or just report the tailored error right there)

File: src/librustc_mir/transform/promote_consts.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs]: Ah I see, thanks.

I think this confirms that somebody else ought to review this, if I cannot even tell as much. ;)

Comment (eddyb) [reply src/librustc_mir/transform/promote_consts.rs]: Almost all of this PR is plumbing to be able to call `is_copy_modulo_regions`, which is the main way you can check for `Copy` outside of the MIR borrowck.

File: src/librustc_mir/transform/check_consts/mod.rs [hunk 3]
```diff
@@ -34,6 +34,7 @@ impl Item<'mir, 'tcx> {
         def_id: DefId,
         body: mir::ReadOnlyBodyAndCache<'mir, 'tcx>,
     ) -> Self {
+        // FIXME: pass down the param env from the caller
```
Comment (eddyb) [inline src/librustc_mir/transform/check_consts/mod.rs]: It should be correct to get it here, no?

Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/mod.rs]: yes, but we also compute it outside, seemed like a good situation to remove an unnecessary query call.

Comment (eddyb) [reply src/librustc_mir/transform/check_consts/mod.rs]: What if the caller got the `ParamEnv` from this `Item` struct instead?

Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/mod.rs]: done

**Comment by oli-obk** [comment]
r? @eddyb 

I split the PR into commits where each commit passes tests. The last commit (the only one actually changing any behavior) is now super small.

**Review by eddyb — APPROVED** [review]
r=me on the changes themselves, but I don't know if we want to land the last commit without some process? I don't remember what the status of this was, did we decide already?

**Comment by oli-obk** [comment]
@Centril does this need an FCP or is an FCP for stabilization at some point enough?

File: src/librustc_mir/transform/promote_consts.rs [hunk 4]
```diff
@@ -118,8 +122,8 @@ impl Candidate {
     /// Returns `true` if we should use the "explicit" rules for promotability for this `Candidate`.
     fn forces_explicit_promotion(&self) -> bool {
```
Comment (RalfJung) [inline src/librustc_mir/transform/promote_consts.rs:123]: As discussed before, could you add a `delay_span_bug` somewhere that ensure that when explicit promotion fails, compilation also fails?

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:123]: You mean https://github.com/rust-lang/rust/blob/bd18bc9a4c142100a7230e1fd1cfe6be08214693/src/librustc_codegen_ssa/mir/operand.rs#L447-L455 :rofl: ?

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:123]: No I don't think so... I mean what we discussed [recently here](https://github.com/rust-lang/rust/pull/70042#issuecomment-600620211).

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:123]: More seriously. We'd need to match on the `DefKind` and probably somehow encode the difference between explicit and implicit promotion: https://github.com/rust-lang/rust/blob/master/src/librustc_mir/const_eval/eval_queries.rs#L354-L357

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:123]: Actually, we could just carry this information in the `mir::Body`,

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:123]: What I would have thought we could do is to change this code here

https://github.com/rust-lang/rust/blob/a4e551ef26e9ba0c2f536e67e6bd33447d8a148a/src/librustc_mir/transform/promote_consts.rs#L743

and add something like
```rust
if validator.explicit && !is_promotable { delay_span_bug(); }
```

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:123]: By "promotion fails" I did not mean "evaluating the promoted fails", I meant "promotion was found not to be possible". We have the key invariant that explicit promotion rules may only be used if, when these rules fail, the program would anyway fail to compile.

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:123]: Good catch!

```
-> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:77:5
   |
LL | /     fn no_impl_copy_empty_value_no_elements() {
LL | |         let x = None;
LL | |         let arr: [Option<Bar>; 0] = [x; 0];
LL | |     }
   | |_____^

error: internal compiler error: Explicit promotion requested, but failed to promote
  --> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:82:5
   |
LL | /     fn no_impl_copy_empty_value_single_element() {
LL | |         let x = None;
LL | |         let arr: [Option<Bar>; 1] = [x; 1];
LL | |     }
   | |_____^

error: internal compiler error: Explicit promotion requested, but failed to promote
  --> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:87:5
   |
LL | /     fn no_impl_copy_value_no_elements() {
LL | |         let x = Some(Bar);
LL | |         let arr: [Option<Bar>; 0] = [x; 0];
LL | |     }
   | |_____^

error: internal compiler error: Explicit promotion requested, but failed to promote
  --> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:92:5
   |
LL | /     fn no_impl_copy_value_single_element() {
LL | |         let x = Some(Bar);
LL | |         let arr: [Option<Bar>; 1] = [x; 1];
LL | |     }
   | |_____^
```

File: src/librustc_mir/transform/promote_consts.rs [hunk 5]
```diff
@@ -207,14 +207,26 @@ impl<'tcx> Visitor<'tcx> for Collector<'_, 'tcx> {
...
                     self.span,
-                ) {
+                );
+                let n = n.try_eval_usize(self.ccx.tcx, self.ccx.param_env);
+                let length_requires_copy_or_promotion = match n {
+                    // Unevaluable (e.g. too generic) -> assume > 1.
+                    None => true,
+                    // A single element does not need copying, we can just use the given element.
+                    // Zero elements don't even need that one element.
+                    Some(0..=1) => false,
```
Comment (RalfJung) [inline src/librustc_mir/transform/promote_consts.rs:224]: Ah, this makes sense. :D

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:224]: It also seems quite fragile (different compiler passes independently making the same decision) but I am not sure if there is a way around that.

File: src/librustc_mir/transform/promote_consts.rs [hunk 6]
```diff
@@ -741,6 +753,14 @@ pub fn validate_candidates(
             // and `#[rustc_args_required_const]` arguments here.
 
             let is_promotable = validator.validate_candidate(candidate).is_ok();
+
+            if validator.explicit && !is_promotable {
+                ccx.tcx.sess.delay_span_bug(
```
Comment (RalfJung) [inline src/librustc_mir/transform/promote_consts.rs:758]: Could you add a comment explaining why this check is absolutely important?
I want to be sure someone doesn't just remove it when it seems convenient.^^

The `explicit` field's doc comment already says a few things, so here's a proposal:
```
// If we use explicit validation, we carry the risk of turning a legitimate run-time
// operation into a failing compile-time operation. Make sure that does not happen
// by asserting that there is no possible run-time behavior here in case promotion fails.
```

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:758]: What actually happens is not that we turn it into a failing compile-time op, it's that we don't promote at all, thus keeping the runtime operation, which happened to be the right thing in the cases that the assertion caught.

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:758]: > keeping the runtime operation

We do that *because we don't even try*. But the branch here is for after we tried, so we cannot be in the case where we want to keep the runtime operation.

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:758]: Yes we are, this checks whether we can promote and if not, we remove it from the list. Then we process this list in `promote_candidates`, which can only fail to promote without ICEing in case of nested promotion candidates, where promoting one candidate eliminated the other candidate because the other candidate is part of the already promoted one.

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:756]: I am entirely confused now.

Going back a step -- did you want to say above that my proposed comment was incorrect, and suggest a better one? I don't understand what is is you are trying to say here.

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:756]: Yes, I believe your comment is incorrect

```
// If we use explicit validation, we expect promotion to succeed.
// If an explicit promotion candidate were to get "unpromoted" here,
// that means we keep an operation that we expect to happen at compile-time
// at runtime. In case of repeat elements, this would just cause an error
// later in compilation (due to multiple moves out of a `!Copy` value), but
// for `rustc_arg_required_const` function arguments, where a constant is
// required, ending up with a runtime value would likely ICE the compiler
// during codegen, or silently cause undocumented and undesired LLVM features
// to be used.
```

**Comment by Centril** [comment]
> @Centril does this need an FCP or is an FCP for stabilization at some point enough?

The PR feels too much in-flux for me to say one way or another. A summary with the rationale, pros, and cons would be helpful.

**Comment by oli-obk** [comment]
This PR changes our behaviour in repeat expressions, turning the repeat element into a full (anonymous) constant if the type is `!Copy` and the repeat count is larger than 1. This means that you can write things like `[Vec::new(); 42]` and don't have to create an intermediate constant yourself. Before this PR you had to write

```rust
const FOO: Vec<i32> = Vec::new();
[FOO; 42]
```

because the repeat element was using the rules for promoteds instead of those for constants.

@scottmcm worries that

>  It seems like it would be bad if adding Copy to the type returned from foo could change what happens in things like [foo(); 4].

Additionally on the tracking issue @RalfJung worried that it may be suprising to users as to what code is accepted and what isn't, since the `Copy` or `const` rule is a first in Rust afaik

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-fail2.rs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// ignore-tidy-linelength
+// ignore-compare-mode-nll
+// compile-flags: -Z borrowck=migrate
+// build-fail
```
Comment (oli-obk) [inline src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-fail2.rs:4]: this is necessary right now, because with `check-pass` the test would succeed. I'm currently looking into it.

Comment (oli-obk) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-fail2.rs:3]: This is a general problem with `--emit=metadata`, we don't compute the `optimized_mir` for functions. It will be fixed when we address that.

File: src/librustc_mir/transform/promote_consts.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:756]: @RalfJung  What do you think about the above comment?

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-fail2.rs (hunk 7 above)
Comment (RalfJung) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-fail2.rs:3]: Is there an issue tracking this?

Comment (oli-obk) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-fail2.rs:3]: I don't believe so and a quick search through our issues didn't show anything. I'll open something...

File: src/librustc_mir/transform/promote_consts.rs (hunk 6 above)
Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:756]: I think we are having different views of the same thing here.

Re: your comment, I am not worried about the case where, if promotion analysis here fails, we ICE the compiler later. I am worried about the case where we do *not* ICE later. Because that means that, no matter whether promotion succeeds or fails, we produce a program, *but the programs may do different things* because we use the less conservative, "explicit promotion" rules. *That* is what I think this comment should explain.

File: src/librustc_mir/transform/promote_consts.rs [hunk 8]
```diff
@@ -200,14 +203,30 @@ impl<'tcx> Visitor<'tcx> for Collector<'_, 'tcx> {
...
+            Rvalue::Repeat(elem, n)
+                if self.ccx.tcx.features().const_in_array_repeat_expressions =>
+            {
+                let is_copy = elem
+                    .ty(&self.ccx.body.local_decls, self.ccx.tcx)
+                    .is_copy_modulo_regions(self.ccx.tcx, self.ccx.param_env, self.span);
+                let n = n.try_eval_usize(self.ccx.tcx, self.ccx.param_env);
+                let length_requires_copy_or_promotion = match n {
+                    // Unevaluable (e.g. too generic) -> assume > 1.
+                    None => true,
```
Comment (RalfJung) [inline src/librustc_mir/transform/promote_consts.rs:219]: So could this lead to a problem if we later figure out that actually, the constant *is* `1`, and then we compile the program after a failed attempt to explicitly promote? (That would trigger the delayed_span_bug now.)

Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs:219]: the alternative is to fail to compile (`!Copy` types treat generic lengths as `> 1`, so if it later happens to be `1` that's irrelevant). While we could permit that, we'd be adding a new source of post monomorphization errors.

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:219]: I mean I am fine with this as long as it is guaranteed that this cannot ICE due to the safety net `delay_span_bug` this PR also adds.
I want to be sure that we do not draw ourselves into a corner where, or backwards compatibility reasons, we have to remove that safety net to fix an ICE.

**Comment by Centril** [comment]
We discussed this at Thursday's language team meeting (2020-04-09). Due to nervousness regarding extending promotion to more places, the rough consensus of those present at the meeting were that we should first consider something like `const { ... }` or `const $expr` (@Centril prefers the former, @nikomatsakis prefers the latter) which has been floated (and I believe @oli-obk has an experimental branch for it) before moving towards accepting e.g., `[Vec::new(); size]`. The suggested `const { ... }` may be introduced experimentally, with a tracking issue, but we'll want an RFC for it (perhaps for stabilization?).

We'll also add a checkbox to the tracking issue https://github.com/rust-lang/rust/issues/49147 to make sure that we consider whether `[const { ... }; size]` has become too much of a pain. But even so, we could possibly relax the rules after stabilizing that feature.

**Comment by RalfJung** [comment]
@oli-obk at least the `delay_span_bug` in this PR is something we want though, IMO... could you submit a follow-up with that part?

File: src/librustc_mir/transform/promote_consts.rs (hunk 6 above)
Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:756]: (This is the `delay_span_bug` that I meant in my other comment, that we should add nevertheless.)

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/71198 is that follow-up. Thanks!
